rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,620 @@
+  It is designed to dynamically store the Sparse Weights(Parameters) of DLRMs.
+  """
+
+  def __init__(self,
+               key_dtype=dtypes.int64,
+               value_dtype=dtypes.float32,
+               dim=1,
+               devices=["/CPU:0", ],
+               partitioner=default_partition_fn,
+               shared_name=None,
+               name="TrainableWrapForDynamicEmbedding",
+               initial_value=0.0,
+               initial_mode="constant",  # use_default=False, == "constant"
+               trainable=True,
+               checkpoint=True):
+
+    """Creates an empty `Variable` object.
+
+    Creates a group of tables placed on devices,
+    the type of its keys and values are specified by key_dtype
+    and value_dtype, respectively.
+
+    Args:
+      key_dtype: the type of the key tensors.
+      value_dtype: the type of the value tensors.
+      dim: the length of the value array for each key.
+      devices: the list of devices holding the tables.
+        One table will be created on each device.
+      partitioner: partition function of keys,
+        return the partition index for each key.
+
+      Example partition func:
+      ```python
+      def default_partition_fn(keys, shard_num):
+        return tf.cast(keys % shard_num, dtype=tf.int32)
+      ```
+
+      shared_name: If non-empty, the SparseVariable will be shared under
+        the given name across multiple sessions.
+      name: A name for the operation (optional).
+      initial_value: The value to use if a key is missing in the hash table.
+      initial_mode: define the behavior when some keys were missing.
+        'random': lookup will return the random values
+          with normal distribution(mean=0.0, stddev=0.1)
+        'constant': lookup will return the valuse
+          filled by `initial_value`.
+      trainable: True, will be treated as a trainable Variable, and add to
+        to the list of variables collected in the graph under the key
+        `GraphKeys.TRAINABLE_VARIABLES`.
+      checkpoint: if True, the contents of the Variable are
+        saved to and restored from checkpoints.
+        If `shared_name` is empty for a checkpointed table,
+        it is shared using the table node name.
+
+    Returns:
+      A `dynamic_embedding.Variable` object.
+    """
+    self.key_dtype = key_dtype
+    self.value_dtype = value_dtype
+    self.default_value = initial_value
+    self.dim = dim
+    self.devices = devices if isinstance(devices, list) else [devices, ]
+    self.partition_fn = partitioner
+    self.name = name
+    self.shared_name = shared_name or "shared_name.{}".format(name)
+    self.initial_value = initial_value
+    self.initial_mode = initial_mode
+    self.use_default = self.initial_mode == 'constant'
+    self.trainable = trainable
+    self.checkpoint = checkpoint
+
+    self._tables = []
+    self.size_ops = []
+    self.shard_num = len(self.devices)
+
+    assert initial_mode in ["random", "constant"], \
```

Comment:
No assert please.

Reply:
> No assert please.
Ok, but why ? I have seen many asserts for checking the args in the core code.